### PR TITLE
Fix for #1154

### LINKED ISSUE #1154: `subs(1/x, x, 0)` is 0
From symengine/SymEngine.jl#22

### PR DESCRIPTION
**Pull request by ranjithkumar007**
When i tried to port SymEngine's pow function to the way SymPy does it ,i found that there is a complementary bug in test_mexprpoly.cpp where x^-1 is evaluated for x as 0.
Until now , It didn't raised (0*Inf) error as 1/0 was evaluated to 0.So I changed that part as well in the same commit.Please review this bug fix and suggest changes. 

Fixes #1154

### CODE REVIEW

File: symengine/pow.cpp [hunk 1]
```diff
@@ -17,7 +17,7 @@ bool Pow::is_canonical(const Basic &base, const Basic &exp) const
 {
     // e.g. 0**x
     if (is_a<Integer>(base) and down_cast<const Integer &>(base).is_zero())
-        return false;
+        return true;
```
Comment (isuruf) [inline symengine/pow.cpp]: This would give true to `0**2` as well. This should match the changes in `pow`

**Comment by certik** [comment]
@ranjithkumar007 thanks for fixing it!

What is going on with the failure here: https://travis-ci.org/symengine/symengine/jobs/189861226 ? Is that a bug in Travis --- should we restart the test?

**Comment by isuruf** [comment]
It's a formatting error. https://travis-ci.org/symengine/symengine/jobs/189861226#L640-L662

**Comment by ranjithkumar007** [comment]
tests have passed now.

File: symengine/tests/basic/test_subs.cpp [hunk 2]
```diff
@@ -152,6 +152,11 @@ TEST_CASE("Mul: subs", "[subs]")
     r1 = mul(x, y)->subs({{x, real_double(0.0)}});
     r2 = real_double(0.0);
     REQUIRE(eq(*r1, *r2));
+
+    d.clear();
+    r1 = div(one, x);
```
Comment (isuruf) [inline symengine/tests/basic/test_subs.cpp]: Can you add a test for `2/x` and `1/(x*y)` as well?

**Comment by isuruf** [comment]
Can you fix your system time? Your git commits have times in the future and this messes up git and github.

File: symengine/tests/basic/test_subs.cpp (hunk 2 above)
Comment (ranjithkumar007) [inline symengine/tests/basic/test_subs.cpp]: @isuruf  in 1/(x*y) case should i add a test with x as 0 and y as 0 [gives output ComplexInf] or only x as 0[gives output ComplexInf/y] ?

**Comment by ranjithkumar007** [comment]
@isuruf  is it ok with git commits now ?

File: symengine/tests/basic/test_arit.cpp [hunk 3]
```diff
@@ -559,6 +559,15 @@ TEST_CASE("Pow: arit", "[arit]")
...
     REQUIRE(eq(*r1, *r2));
 
+    r1 = pow(zero, zero);
+    REQUIRE(eq(*r1, *one));
+
+    r1 = pow(zero, i2);
+    REQUIRE(eq(*r1, *zero));
+
+    r1 = pow(zero, im1);
+    REQUIRE(r1->__str__() == "zoo");
```
Comment (isuruf) [inline symengine/tests/basic/test_arit.cpp]: Can you change this to `eq(*r1, *ComplexInf)`? (You might need to add `using SymEngine::ComplexInf` to the top

File: symengine/tests/basic/test_subs.cpp [hunk 4]
```diff
@@ -152,6 +152,21 @@ TEST_CASE("Mul: subs", "[subs]")
...
+    d[x] = zero;
+    REQUIRE((*r1->subs(d)).__str__() == "zoo");
```
Comment (isuruf) [inline symengine/tests/basic/test_subs.cpp]: Same here

File: symengine/tests/basic/test_subs.cpp [hunk 5]
```diff
@@ -152,6 +152,21 @@ TEST_CASE("Mul: subs", "[subs]")
...
+
+    d.clear();
+    r1 = div(i2, x);
+    d[x] = zero;
+    REQUIRE((*r1->subs(d)).__str__() == "zoo");
+
+    d.clear();
+    r1 = div(one, mul(x, y));
+    d[x] = zero;
+    REQUIRE((*r1->subs(d)).__str__() == "zoo/y");
```
Comment (isuruf) [inline symengine/tests/basic/test_subs.cpp]: And here to `eq(*r1, *div(ComplexInf, y))`

**Comment by isuruf** [comment]
@ranjithkumar007, thanks for the PR.

File: symengine/pow.cpp [hunk 6]
```diff
@@ -89,8 +94,19 @@ RCP<const Basic> pow(const RCP<const Basic> &a, const RCP<const Basic> &b)
...
+
+    if (eq(*a, *zero)) {
+        if (is_a_Number(*b)
+            and rcp_static_cast<const Number>(b)->is_positive()) {
+            return zero;
+        } else if (is_a_Number(*b)
+                   and rcp_static_cast<const Number>(b)->is_negative()) {
+            return ComplexInf;
+        } else {
+            return make_rcp<const Pow>(a, b);
```
Comment (isuruf) [inline symengine/pow.cpp]: I don't know whether this is desired. In SymPy, assumptions can check whether `b > 0` or not. In symengine we cannot and therefore keeps `0**sqrt(2)` as it is.

**Pull request merged**